pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java

```diff
@@ -4,15 +4,144 @@
+import net.sourceforge.pmd.lang.java.ast.ASTAnyTypeDeclaration.TypeKind;
 import net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTConstructorDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTEnumDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTInitializer;
+import net.sourceforge.pmd.lang.java.ast.AccessNode;
 import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+import net.sourceforge.pmd.properties.RegexProperty;
+import net.sourceforge.pmd.properties.RegexProperty.RegexPBuilder;
+
 
+/**
+ * Configurable naming conventions for type declarations.
+ */
 public class ClassNamingConventionsRule extends AbstractJavaRule {
 
+    private static final RegexProperty CLASS_REGEX = defaultProp("class").desc("Regex which applies to concrete class names").build();
+    private static final RegexProperty ABSTRACT_CLASS_REGEX = defaultProp("abstract class").build();
+    private static final RegexProperty INTERFACE_REGEX = defaultProp("interface").build();
+    private static final RegexProperty ENUMERATION_REGEX = defaultProp("enum").build();
+    private static final RegexProperty ANNOTATION_REGEX = defaultProp("annotation").build();
+    private static final RegexProperty UTILITY_CLASS_REGEX = defaultProp("utility class").defaultValue("[A-Z][a-zA-Z]+Util").build();
+
+
+    public ClassNamingConventionsRule() {
+        definePropertyDescriptor(CLASS_REGEX);
+        definePropertyDescriptor(ABSTRACT_CLASS_REGEX);
+        definePropertyDescriptor(INTERFACE_REGEX);
+        definePropertyDescriptor(ENUMERATION_REGEX);
+        definePropertyDescriptor(ANNOTATION_REGEX);
+        definePropertyDescriptor(UTILITY_CLASS_REGEX);
+
+        addRuleChainVisit(ASTClassOrInterfaceDeclaration.class);
+        addRuleChainVisit(ASTEnumDeclaration.class);
+        addRuleChainVisit(ASTAnnotationTypeDeclaration.class);
+    }
+
+
+    private void checkMatches(ASTAnyTypeDeclaration node, PropertyDescriptor<Pattern> regex, Object data) {
+        if (!getProperty(regex).matcher(node.getImage()).matches()) {
+            addViolation(data, node, new Object[]{
+                    isUtilityClass(node) ? "utility class" : node.getTypeKind().getPrintableName(),
+                    node.getImage(),
+                    getProperty(regex).toString(),
+            });
+        }
+    }
+
+
+    // This could probably be moved to ClassOrInterfaceDeclaration
+    // to share the implementation and be used from XPath
+    private boolean isUtilityClass(ASTAnyTypeDeclaration node) {
+        if (node.getTypeKind() != TypeKind.CLASS) {
+            return false;
+        }
+
+        // A class without declarations shouldn't be reported
+        boolean hasAny = false;
+
+        // we could probably enrich ASTAnyTypeBodyDeclaration
+        // with methods to know what kind of declaration it is
```

Comment:
I totally support this. Doing so would be feasible without breaking APIs, so feel free to do so